### ISSUE TITLE
java.lang.ClassNotFoundException: AnnotationDatabaseImpl

### ISSUE THREAD

**Issue by elioth010** [roboguice/roboguice#246]
With RoboGuice 3.0 launch this exception:

10-28 11:48:51.601: W/dalvikvm(24040): threadid=1: thread exiting with uncaught exception (group=0x417e8c08)
10-28 11:48:51.606: E/AndroidRuntime(24040): FATAL EXCEPTION: main
10-28 11:48:51.606: E/AndroidRuntime(24040): Process: com.bytesw.perfil, PID: 24040
10-28 11:48:51.606: E/AndroidRuntime(24040): java.lang.RuntimeException: Unable to start activity ComponentInfo{com.bytesw.perfil/com.bytesw.perfil.view.activity.AutenticacionActivity}: java.lang.IllegalStateException: Unable to use annotation database(s)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2305)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2363)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at android.app.ActivityThread.access$900(ActivityThread.java:161)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1265)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at android.os.Handler.dispatchMessage(Handler.java:102)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at android.os.Looper.loop(Looper.java:157)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at android.app.ActivityThread.main(ActivityThread.java:5356)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at java.lang.reflect.Method.invokeNative(Native Method)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at java.lang.reflect.Method.invoke(Method.java:515)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1265)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1081)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at de.robv.android.xposed.XposedBridge.main(XposedBridge.java:126)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at dalvik.system.NativeStart.main(Native Method)
10-28 11:48:51.606: E/AndroidRuntime(24040): Caused by: java.lang.IllegalStateException: Unable to use annotation database(s)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at roboguice.RoboGuice.initializeAnnotationDatabaseFinderAndHierarchyTraversalFilterFactory(RoboGuice.java:292)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at roboguice.RoboGuice.getOrCreateBaseApplicationInjector(RoboGuice.java:149)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at roboguice.RoboGuice.getOrCreateBaseApplicationInjector(RoboGuice.java:90)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at roboguice.RoboGuice.getInjector(RoboGuice.java:197)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at roboguice.activity.RoboActionBarActivity.onCreate(RoboActionBarActivity.java:85)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at com.bytesw.perfil.view.activity.AutenticacionActivity.onCreate(AutenticacionActivity.java:26)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at android.app.Activity.performCreate(Activity.java:5426)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1105)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2269)
10-28 11:48:51.606: E/AndroidRuntime(24040):    ... 12 more
10-28 11:48:51.606: E/AndroidRuntime(24040): Caused by: java.lang.RuntimeException: java.lang.ClassNotFoundException: AnnotationDatabaseImpl
10-28 11:48:51.606: E/AndroidRuntime(24040):    at com.google.inject.AnnotationDatabaseFinder.<init>(AnnotationDatabaseFinder.java:47)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at com.google.inject.Guice.setAnnotationDatabasePackageNames(Guice.java:133)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at roboguice.RoboGuice.initializeAnnotationDatabaseFinderAndHierarchyTraversalFilterFactory(RoboGuice.java:288)
10-28 11:48:51.606: E/AndroidRuntime(24040):    ... 20 more
10-28 11:48:51.606: E/AndroidRuntime(24040): Caused by: java.lang.ClassNotFoundException: AnnotationDatabaseImpl
10-28 11:48:51.606: E/AndroidRuntime(24040):    at java.lang.Class.classForName(Native Method)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at java.lang.Class.forName(Class.java:251)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at java.lang.Class.forName(Class.java:216)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at com.google.inject.AnnotationDatabaseFinder.getAnnotationDatabaseInstance(AnnotationDatabaseFinder.java:73)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at com.google.inject.AnnotationDatabaseFinder.<init>(AnnotationDatabaseFinder.java:36)
10-28 11:48:51.606: E/AndroidRuntime(24040):    ... 22 more
10-28 11:48:51.606: E/AndroidRuntime(24040): Caused by: java.lang.NoClassDefFoundError: AnnotationDatabaseImpl
10-28 11:48:51.606: E/AndroidRuntime(24040):    ... 27 more
10-28 11:48:51.606: E/AndroidRuntime(24040): Caused by: java.lang.ClassNotFoundException: Didn't find class "AnnotationDatabaseImpl" on path: DexPathList[[zip file "/data/app/com.bytesw.perfil-1.apk"],nativeLibraryDirectories=[/data/app-lib/com.bytesw.perfil-1, /vendor/lib, /system/lib]]
10-28 11:48:51.606: E/AndroidRuntime(24040):    at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:67)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at java.lang.ClassLoader.loadClass(ClassLoader.java:497)
10-28 11:48:51.606: E/AndroidRuntime(24040):    at java.lang.ClassLoader.loadClass(ClassLoader.java:457)
10-28 11:48:51.606: E/AndroidRuntime(24040):    ... 27 more

Activity
<code>
@ContentView(R.layout.activity_autenticacion)  
public class AutenticacionActivity extends AbstractActivity {  
    @InjectView(R.id.edit_username) private EditText userTextView;  
    @InjectView (R.id.edit_password) private EditText passwordTextView;  
    @Override  
    protected void onCreate(Bundle savedInstanceState) {  
        super.onCreate(savedInstanceState);  
    }  
}  
</code>
Abstract:
<code>
import roboguice.activity.RoboActionBarActivity;  
public abstract class AbstractActivity extends RoboActionBarActivity{  
    public abstract void show(final String message);  
    public abstract void handleException(final Exception e);  
}  
</code>

**Comment by enil** [comment]
I got the same exception as well, but it seem to be caused by not having any actual fields or methods annotated with `@Inject` or `@InjectView`. Once I added an injected field the problem went away (no, it doesn't make a lot of sense not to have any injected fields, but for a beginner who just want everything to compile before trying it out the error is very baffling).

I'm not very knowledgable about the annotation preprocessor, but it seems to me like it's never run unless any of those annotations are used, thus `AnnotationDatabaseImpl` is never generated.

**Comment by ViswanathL** [comment]
I was also facing this issue. I made this working by adding javax.inject-1.jar

**Comment by akiwarheit** [comment]
Already have the ff. dependencies
```
    compile 'org.roboguice:roboguice:3.0.1'
    provided 'org.roboguice:roboblender:3.0'
```
but I'm still getting the error.

**Comment by sasij** [comment]
Hi!


I have problems with Roboguice 3.0.1 when I use proguard. If i disable proguard, my code run OK. I read the old entries and disabled roboblender in my Application class:

``` java
static { RoboGuice.setUseAnnotationDatabases(false);}
```
but now (with proguard enable) it seem as if it unable to create injector or needs that i have to code a constructor always.Without proguard, it uses the default constructor and my code run well. I get the next error:

```
   java.lang.RuntimeException: Unable to start activity ComponentInfo{com.peerade.ave/com.peerade.ave.views.activities.HomeActivity}: com.google.inject.ConfigurationException: Guice configuration errors:
    1) Could not find a suitable constructor in com.peerade.ave.presenters.HomeFragmentPresenter. Classes must have either one (and only one) constructor annotated with @Inject or a zero-argument constructor that is not private.
            at com.peerade.ave.presenters.HomeFragmentPresenter.class(Unknown Source)
    while locating com.peerade.ave.presenters.HomeFragmentPresenter
    for field at com.peerade.ave.views.fragments.HomeFragment.presenter(Unknown Source)
    while locating com.peerade.ave.views.fragments.HomeFragment
    1 error
            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2298)
            at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2360)
            at android.app.ActivityThread.access$800(ActivityThread.java:144)
            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1278)
            at android.os.Handler.dispatchMessage(Handler.java:102)
            at android.os.Looper.loop(Looper.java:135)
            at android.app.ActivityThread.main(ActivityThread.java:5221)
            at java.lang.reflect.Method.invoke(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:372)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:899)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:694)
```

I'm using Android Studio 1.0.2 and my proguard file contain:

```
-target 1.6
-dontobfuscate
-dontoptimize
-dontusemixedcaseclassnames
-dontskipnonpubliclibraryclasses
-dontpreverify
-verbose


# The -optimizations option disables some arithmetic simplifications that Dalvik 1.0 and 1.5 can't handle.
-optimizations !code/simplification/arithmetic

-keep public class * extends android.app.Activity
-keep public class * extends android.app.Application
-keep public class * extends android.app.Service
-keep public class * extends android.content.BroadcastReceiver
-keep public class * extends android.content.ContentProvider
-keep class com.google.inject.Binder
-keepclassmembers class * {
    @com.google.inject.Inject <fields>;
    @com.google.inject.Inject <init>(...);
}
# There's no way to keep all @Observes methods, so use the On*Event convention to identify event handlers
-keepclassmembers class * {
    void *(**On*Event);
}
-keep public class * extends android.view.View {
    public <init>(android.content.Context);
    public <init>(android.content.Context, android.util.AttributeSet);
    public <init>(android.content.Context, android.util.AttributeSet, int);
    public void set*(...);
}
-keep public class roboguice.**
-keep class com.peerade.ave.roboguice.modules.AveModule

##Retrofit
-keep class com.google.gson.** { *; }
-keep class com.google.inject.** { *; }
-keep class org.apache.http.** { *; }
-keep class org.apache.james.mime4j.** { *; }
-keep class javax.inject.** { *; }

-keepattributes *Annotation*
-keepattributes Signature
-keep class com.squareup.okhttp.** { *; }
-keep interface com.squareup.okhttp.** { *; }
-keep class retrofit.** { *; }
-keepclasseswithmembers class * {
    @retrofit.http.* <methods>;
}

## Smack
-keep class * implements org.jivesoftware.smackx.xevent.MessageEventNotificationListener
-keep class * extends org.jivesoftware.smackx.DefaultMessageEventRequestListener
-keep class org.jivesoftware.smackx.xevent.** { *; }
-keep interface org.jivesoftware.smackx.xevent.** { *; }


## Warning
-dontwarn okio.**
-dontwarn roboguice.**
-dontwarn org.roboguice.**
-dontwarn retrofit.**
-dontwarn com.google.common.**
-dontwarn org.xbill.**
-dontwarn com.squareup.okhttp.**
-dontwarn com.nhaarman.**
-dontwarn com.facebook.android.BuildConfig

-keepclassmembers class * extends java.lang.Enum {
    public static **[] values();
    public static ** valueOf(java.lang.String);
}

-keep class * implements android.os.Parcelable {
  public static final android.os.Parcelable$Creator *;
}

# Removes all calls to Log. Delete the methods you want to keep.
-assumenosideeffects class android.util.Log {
    public static int v(...);
    public static int d(...);
    public static int i(...);
    public static int w(...);
    public static int e(...);
    public static int wtf(...);
}

```

Any help please? 
Thanks a lot.

**Comment by eaorak** [comment]
I have faced with the same problem on Eclipse environment and no suggestions worked so far. I have added meta-data information to AndroidManifest.xml, disabled useAnnotationDatabases on Application class, added roboblender as a library but no luck. Any help ?

**Comment by Maddoc42** [comment]
@sasij I had the same problem today. Try excluding all your classes and methods from proguard (e.g. ``-keep class <my.package>.** { *; }``).

In some cases the ClassNotFoundException (AnnotationDatabaseImpl) seems to originate from proguard as well.

**Comment by sasij** [comment]
@Maddoc42 Many Thanks. I will try it. I had to continue with my app after creating all my constructors.

**Comment by ZetBrush** [comment]
I also solved the problem by adding
static {
        RoboGuice.setUseAnnotationDatabases(false);
    }

v3.0.1.jar

**Comment by ForerverSoul** [comment]
@enil OMG, you're the genius. 
If there's no any fields or methods annotated with the @Inject then the exception will be thrown.
That's it! And this is definitely a bug.
Maybe this is why all those people who followed the wiki and still can't solve this problem.
Thank you @enil .
And to @stephanenicolas , I think you should fix this issue as soon as possible.

**Comment by enil** [comment]
@ForerverSoul I think this is because how the annotation processing in Java works, if no annotation of the type is found it is never invoked as far as I know. I'm not sure that's something that can be fixed. Maybe a note in the documentation somewhere could be helpful however.

**Comment by ForerverSoul** [comment]
@enil I think there are something will process the annotations and generate related AnnotationDatabaseImpl.class. If it is, then it can be detected if there is nothing to do.
Just skip loading this class after the generation be ignored by no any annotation.
But a note in the documentation is enough though.

**Comment by asco33** [comment]
Is there any news on this? 

I tried adding a 'dummy' injected view to the BaseActivity and BaseFragment classes that all my UI classes derive from like 

    @Nullable
    @InjectView(R.id.drawer_layout)
    View dummy;

which seemed to slightly reduce the crashes. 

My next step will be to try ZetBrush's suggestion. 

If that doesn't work I would have to completely remove RG from a vast project because for a small percentage of users this bug makes the app unusable who then proceed to create a sh**storm on the Google Play reviews. 

Android Studio & Gradle, using

    compile 'org.roboguice:roboguice:3.+'
    provided 'org.roboguice:roboblender:3.+'
    compile 'com.google.code.findbugs:jsr305:1.3.9'

**Comment by heatherSnepenger** [comment]
I'm having this issue as well. I've got this in my gradle file:

      compile 'org.roboguice:roboguice:3.+'
      provided 'org.roboguice:roboblender:3.+'

and I've tried building with api 21 and 22. I've tried removing `provided 'org.roboguice:roboblender:3.+'`, and I've tried adding 

    static {
         RoboGuice.setUseAnnotationDatabases(false);
    }

Anyone know what else could be the problem?

**Comment by OlegDokuka** [comment]
Hi guys, have anybody tried to use roboguice with groovy on android, i got the same issue when I just run my application 

my grade 
``` groovy 
apply plugin: 'com.android.application'
apply plugin: 'groovyx.grooid.groovy-android'

android {
    compileSdkVersion 22
    buildToolsVersion "21.1.2"
    defaultConfig {
        applicationId "ua.rainbow.development.clickwalker"
        minSdkVersion 14
        targetSdkVersion 22
        versionCode 1
        versionName "1.0"
    }
    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
        }
    }
    productFlavors {
    }
}

dependencies {
    compile fileTree(dir: 'libs', include: ['*.jar'])
    compile 'com.android.support:appcompat-v7:22.1.1'
    compile 'io.reactivex:rxandroid:0.24.0'
    compile 'org.roboguice:roboguice:3.0.1'
    provided 'org.roboguice:roboblender:3.0.1'
    //compile 'io.reactivex:rxjava-async-util:0.21.0'
    //compile 'com.squareup.dagger:dagger:1.2.2'
    compile 'org.codehaus.groovy:groovy:2.4.3:grooid'
    // provided 'com.squareup.dagger:dagger-compiler:1.2.2'
}
```

my groovy activity
```groovy
package ua.rainbow.development.clickwalker

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.TextView
import com.google.inject.Inject
import roboguice.RoboGuice
import roboguice.activity.RoboActivity
import roboguice.inject.ContentView
import roboguice.inject.InjectView
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import ua.rainbow.development.clickwalker.service.UserService

/**
 *  Created by shadowgun on 27.05.15.
 *  email: $ { EMAIL }*/
@ContentView(R.layout.activity_login)
class LoginActivity extends RoboActivity {
    static {
        RoboGuice.setUseAnnotationDatabases(false);
    }
    @Inject
    private UserService userService;
    @InjectView(R.id.hello_world)
    private TextView helloWorld;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        userService.user.subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnNext({
            helloWorld.text = it;
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_login, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        def id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }
}
```

my manifest

``` xml
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    package="ua.rainbow.development.clickwalker">

    <meta-data
        android:name="ua.rainbow.development.clickwalker.module"
        android:value="ua.rainbow.development.clickwalker.module.UserModule" />
    <meta-data
        android:name="roboguice.annotations.packages"
        android:value="ua.rainbow.development.clickwalker"/>
    <application
        android:allowBackup="true"
        android:icon="@mipmap/ic_launcher"
        android:label="@string/app_name"
        android:theme="@style/AppTheme">
        <activity
            android:name="ua.rainbow.development.clickwalker.LoginActivity"
            android:label="@string/app_name">
            <intent-filter>
                <action android:name="android.intent.action.MAIN" />

                <category android:name="android.intent.category.LAUNCHER" />
            </intent-filter>
        </activity>
    </application>

</manifest>
```

my program file
```
# Add project specific ProGuard rules here.
# By default, the flags in this file are appended to flags specified
# in /Users/shadowgun/Library/Android/sdk/tools/proguard/proguard-android.txt
# You can edit the include path and order by changing the proguardFiles
# directive in build.gradle.
#
# For more details, see
#   http://developer.android.com/guide/developing/tools/proguard.html

# Add any project specific keep options here:

# If your project uses WebView with JS, uncomment the following
# and specify the fully qualified class name to the JavaScript interface
# class:
#-keepclassmembers class fqcn.of.javascript.interface.for.webview {
#   public *;
#}
-dontobfuscate
-keep public class AnnotationDatabaseImpl
```
I run my project in android studio

**Comment by OlegDokuka** [comment]
solved. I`v just recreated project and it start working. It is very strange

**Comment by ghostofcain** [comment]
The only thing that worked for me was the suggestion by ZetBrush (i.e. disabling the AnnotationDatabases).

**Comment by grgaaa** [comment]
One more information for developers. In my case the error started to appear after adding play-services dependencies to my project.

This means I added the following to the gradle file:

- compile 'com.google.android.gms:play-services:7.8.0'
- multiDexEnabled = true
- dexOptions {javaMaxHeapSize "4g"}

Using roboguice and roboblender 3.0.1.

**Comment by dmitry-zaitsev** [comment]
Easily reproducable as for now.

1. Create new Android studio project (everything by default, no Play Services, just support libraries).
2. Add RoboGuice + RoboBlender as per documentation
3. Run. Crashes with same error.

I think issue has to be reopened.

**Comment by zawadz88** [comment]
I had the same error, but only when using ProGuard for some reason. My configuration consisted of build tools 23.0.2, Android Gradle plugin 2.0.0-alpha3 and lots of included libraries so multidex was also enabled.
I was able to solve it by adding in the ProGuard config file:

    -keep public class * extends com.google.inject.AnnotationDatabase

And in app's build.gradle:
    
    project.tasks.withType(JavaCompile) { task ->
        options.compilerArgs << "-AguiceAnnotationDatabasePackageName=com.myapp.package"
    }
And finally by adding in the AndroidManifest.xml under application tag:

    <meta-data android:name="roboguice.annotations.packages" android:value="com.myapp.package"/>

I got the idea from: http://www.bhyuu.com/roboblender-use-annotation-databases-with-multiple-modules/

**Comment by arvydaszenitech** [comment]
I had the same issue when added RoboGuice to my project.
I used  this line to inject members: 

```
RoboGuice.getInjector(getApplicationContext()).injectMembers(this);
```
But I wasn't actually injecting anything using @Inject annotation.
Issue was fixed when I injected SharedPreferences
```
@Inject
SharedPreferences sharedPreferences. 
```
Looks like it triggered to use AnnotationDatabase. Then I removed 
SharedPreferences object and it still worked. Not sure what happens under the hood, but its a strange behaviour.
